lib/SendStream.js

```diff
@@ -226,6 +224,74 @@ SendStream.prototype.isConditionalGET = function isConditionalGET () {
     this.req.headers['if-modified-since']
 }
 
+SendStream.prototype.isNotModifiedFailure = function isNotModifiedFailure () {
+  const req = this.req
+  const res = this.res
+
+  // Always return stale when Cache-Control: no-cache
+  // to support end-to-end reload requests
+  // https://tools.ietf.org/html/rfc2616#section-14.9.4
+  if (
+    'cache-control' in req.headers &&
+    req.headers['cache-control'].indexOf('no-cache') !== -1
+  ) {
+    return false
+  }
+
+  // if-none-match
+  if ('if-none-match' in req.headers) {
+    const ifNoneMatch = req.headers['if-none-match']
+
+    if (ifNoneMatch === '*') {
+      return true
+    } else {
+      const etag = res.getHeader('etag')
+
+      if (typeof etag !== 'string') {
+        return false
+      }
+
+      const etagL = etag.length
+      const isMatching = parseTokenList(ifNoneMatch, function (match) {
+        const mL = match.length
```

Comment:
can we do external pure function instead? inline ones cause deoptimization, AFAIR

Reply:
How should i inject etag variable into that external pure function?